web3/eth/base_eth.py

```diff
@@ -95,17 +95,28 @@ def set_gas_price_strategy(
         self._gas_price_strategy = gas_price_strategy
 
     def estimate_gas_munger(
-        self, transaction: TxParams, block_identifier: Optional[BlockIdentifier] = None
-    ) -> Sequence[Union[TxParams, BlockIdentifier]]:
+        self,
+        transaction: TxParams,
+        block_identifier: Optional[BlockIdentifier] = None,
+        state_override: Optional[CallOverride] = None,
+    ) -> Sequence[Union[TxParams, BlockIdentifier, CallOverride]]:
         if "from" not in transaction and is_checksum_address(self.default_account):
             transaction = assoc(transaction, "from", self.default_account)
 
         if block_identifier is None:
```

Comment:
I don't actually think this logic is doing what we want. The `eth_call` munger will set the block to the default block but geth also will use the "pending" block if the block identifier is missing... but even if it is missing, we want to be able to send a `state_override`. It doesn't look like we allow for that with this `return` here.

We should probably abstract the logic of the `call_munger` out and use it here as well. Instead of renaming it let's keep `estimate_gas_munger()` and `call_munger()` separated and let's abstract only one logic block into a `_eth_call_and_estimate_gas_munger()` and use that in each. This way we are not trying to recreate the same logic in both places and possibly behaving one way in one and another way in the other (as it currently appears to be happening).

```py
def _eth_call_and_estimate_gas_munger(...):
    # eth_call munger logic here since that's already written out and tested
    
def call_munger(...):
   return  _eth_call_and_estimate_gas_munger(...)
   
def estimate_gas_munger(...):
   return  _eth_call_and_estimate_gas_munger(...)
```

That way if they ever don't have parity we can still tweak, etc... but for now they share the same logic so let's reduce the surface for friction / errors to occur. Thoughts on that?

Reply:
Oh yep, good catch there. Not sure why I thought returning just the transaction would work, will fix!